molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java

```diff
@@ -0,0 +1,340 @@
+import static org.molgenis.data.security.audit.AuthenticationUtils.isRunByUser;
+import static org.molgenis.data.transaction.TransactionConstants.TRANSACTION_ID_RESOURCE_NAME;
+
+import com.google.common.collect.ForwardingIterator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import org.molgenis.audit.AuditEventPublisher;
+import org.molgenis.data.AbstractRepositoryDecorator;
+import org.molgenis.data.Entity;
+import org.molgenis.data.Fetch;
+import org.molgenis.data.Query;
+import org.molgenis.data.Repository;
+import org.molgenis.data.aggregation.AggregateQuery;
+import org.molgenis.data.aggregation.AggregateResult;
+import org.molgenis.data.util.EntityTypeUtils;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+
+/**
+ * Publishes audit events for CRUD actions. In the case of system entity types, does not publish
+ * events for read actions. If the current user is SYSTEM (excluding elevated users running as
+ * system), does not publish anything.
+ */
+public class AuditingRepositoryDecorator extends AbstractRepositoryDecorator<Entity> {
+
+  static final String ENTITY_READ = "ENTITY_READ";
+  static final String ENTITY_CREATED = "ENTITY_CREATED";
+  static final String ENTITY_UPDATED = "ENTITY_UPDATED";
+  static final String ENTITY_DELETED = "ENTITY_DELETED";
+  static final String ENTITIES_READ = "ENTITIES_READ";
+  static final String ENTITIES_COUNTED = "ENTITIES_COUNTED";
+  static final String ENTITIES_AGGREGATED = "ENTITIES_AGGREGATED";
+  static final String ALL_ENTITIES_DELETED = "ALL_ENTITIES_DELETED";
+  private static final String ENTITY_TYPE_ID = "entityTypeId";
+  private static final String ENTITY_ID = "entityId";
+  private static final String ENTITY_IDS = "entityIds";
+  private static final String RUN_AS = "runAs";
+
+  private final AuditEventPublisher auditEventPublisher;
+
+  public AuditingRepositoryDecorator(
+      Repository<Entity> delegateRepository, AuditEventPublisher auditEventPublisher) {
+    super(delegateRepository);
+    this.auditEventPublisher = requireNonNull(auditEventPublisher);
+  }
+
+  @Override
+  public @Nonnull
+  Iterator<Entity> iterator() {
+    if (isRunByUser() && isNonSystemEntityType()) {
+      return new AuditingIterator(delegate().iterator());
+    } else {
+      return delegate().iterator();
+    }
+  }
+
+  @Override
+  public void forEachBatched(Fetch fetch, Consumer<List<Entity>> consumer, int batchSize) {
+    if (isRunByUser() && isNonSystemEntityType()) {
+      delegate()
+          .forEachBatched(
+              fetch,
+              entities -> {
+                consumer.accept(entities);
+                var ids = entities.stream().map(Entity::getIdValue).collect(toList());
+                audit(ENTITIES_READ, ENTITY_IDS, ids);
```

Comment:
this means the logging happens in batches

Reply:
Correct! :)